fixed bugs in mixed typography

DearWWWWWT opened the pull request.

DearWWWWWT committed:
delete no use files

DearWWWWWT committed:
add other code for output format compare

DearWWWWWT committed:
fixed bugs in mixed typography

1. add a class use for mark character(letter, punctuation, symbol)'s language unit which in a mix-language text
2. edit the logic in mixed language text typography using language unit mark

DearWWWWWT committed:
edit the logic in pure English, now Force Fix can be checked anytime

DearWWWWWT committed:
code formatting

DearWWWWWT committed:
fix normal prpunctuation and symbols position set in mixed typographer

nistream reviewed (changes requested):
1. 请将此 Repo 公开级别改为 SeedV private
2. 请将上级工程 ( splash-renderer-story ) 的集成修改一并提交 pull request
3. 以前之所以采用 submodule 方式集成, 是因为可以直接在上级工程中修改子模块并提交 pr, 若采用 git URL 方式集成, 鉴于 Assets 等基础资源已经在本 repo 中被删除,后续如何进行修改维护?

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+﻿using System.Linq;
```
版权声明照此处理 (适用于所有文件):
1. 对于没有修改的原始文件, 保留原作者的版权声明不变
2. 对于新增的文件, 增加 SeedV 的版权声明
3. 对于修改的文件, 保留原作者的版权声明, 在下方增加 SeedV 的版权声明, 例如:
```
// <Original copyright statement by the previous author>
//
// The SeedV Lab (Beijing SeedV Technology Co., Ltd.) modifications: Copyright 2024 The SeedV Lab
// (Beijing SeedV Technology Co., Ltd.) All Rights Reserved. The modifications in this file are the intellectual property of the SeedV Lab.
```

nistream commented on UPMPackage/Scripts/Runtime/Types/MirroredCharsMaper.cs:
```diff
@@ -0,0 +1,54 @@
+namespace RTLTMPro {
+  public class MirroredCharsMaper {
+    public static readonly Dictionary<char, char> MirroredCharsMap = new Dictionary<char, char>() {
+      // ['('] = ')',
+      // [')'] = '(',
+      // ['»'] = '«',
+      // ['«'] = '»',
```
- 如果非必要请删除
- 如果是有意保留, 注释写明保留的原因

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+
+      #endregion
+
+      //if no Arabic and English character is find, set to English
```
注释规则参考:
[非代码文本规范](https://seedvlab.feishu.cn/wiki/wikcn3rAB43mKMSdGDjDred3N9r#AaGjd22PtoiujQxVSxScIgCznzh) (审议中, 可共同探讨是否合理)

nistream replied:
```suggestion
      // If no Arabic and English character is find, set to English
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+
+          ContextType previousType = ContextType.Default;
+          ContextType behindType = ContextType.Default;
+          // White Space didn't affect text composing in current logic
+          // reserve for farther logic optimization
```
```suggestion
          // White Spaces don't affect text composing in the current logic.
          // Reserved for farther logic optimization.
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+          bool previousWhiteSpace = false;
+          bool behindWhiteSpace = false;
+
+          // search backward for Context.Type
+          // maybe a failed searching 
```
```suggestion
          // Search backward for Context.Type.
          // May be a failed searching. 
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+            }
+          }
+
+          // search forward for Context.Type
+          // maybe a failed searching
```
```suggestion
          // Search forward for Context.Type.
          // May be a failed searching.
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+                    }
+                  }
+
+                  //if right character is rightest, case previous type only
```
```suggestion
                  // If the current character is the last in the input, only consider the previous type.
```
我不确定是不是这个意思？

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+
+                  //if right character is rightest, case previous type only
+                  if (i + j == input.Length - 1) behindType = mirrorPreviousType;
+                  //if previous type is default, there is no letter in or front this mirror character
```
```suggestion
                  // If the previous type is default, assume no letter exists before this character.
```
不确定, 请核对

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+                  if (i + j == input.Length - 1) behindType = mirrorPreviousType;
+                  //if previous type is default, there is no letter in or front this mirror character
+                  if (mirrorPreviousType == ContextType.Default) mirrorPreviousType = mirrorBehindType;
+                  //if all type is default, all text is not letter
```
```suggestion
                  // If both previous and next types are default, assume the text contains no letters.
```
不确定, 请核对

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+                  }
+
+                  //if right character is rightest, case previous type only
+                  if (i + j == input.Length - 1) behindType = mirrorPreviousType;
```
```suggestion
                  if (i + j == input.Length - 1) {
                    behindType = mirrorPreviousType;
                  }
```
A maximum of one statement per line. According to [C# at Google Style Guide](https://google.github.io/styleguide/csharp-style.html)

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+                  //if right character is rightest, case previous type only
+                  if (i + j == input.Length - 1) behindType = mirrorPreviousType;
+                  //if previous type is default, there is no letter in or front this mirror character
+                  if (mirrorPreviousType == ContextType.Default) mirrorPreviousType = mirrorBehindType;
```
```suggestion
                  if (mirrorPreviousType == ContextType.Default) {
                    mirrorPreviousType = mirrorBehindType;
                  }
```
A maximum of one statement per line. According to [C# at Google Style Guide](https://google.github.io/styleguide/csharp-style.html)

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+            isRtl[i] = ContextType.Arabic;
+            continue;
+          }
+          // in order to clearly see the judgment logic, ignore the grammar prompts here
```
```suggestion
          // In order to clearly see the judgment logic, ignore the grammar prompts here
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+            isRtl[i] = previousType;
+            continue;
+          }
+          // in order to clearly see the judgment logic, ignore the grammar prompts here
```
```suggestion
          // In order to clearly see the judgment logic, ignore the grammar prompts here
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+            isRtl[i] = behindType;
+            continue;
+          }
+          // in order to clearly see the judgment logic, ignore the grammar prompts here
```
```suggestion
          // In order to clearly see the judgment logic, ignore the grammar prompts here
```

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,177 @@
+          if (previousWhiteSpace == false && behindWhiteSpace == false) {
+            isRtl[i] = ContextType.Arabic;
+          }
+          // in order to clearly see the judgment logic, ignore the grammar prompts here
```
```suggestion
          // In order to clearly see the judgment logic, ignore the grammar prompts here
```

nistream commented on UPMPackage/Scripts/Runtime/Types/ContextType.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace RTLTMPro
+{
+    public enum ContextType
+    {
+        Default = 0,
+        English = 1,
+        Arabic = 2,
+    }
+}
```
- Indentation of 2 spaces, no tabs.
- No line break before opening brace.

DearWWWWWT committed:
edit the typographer logic for white space and paired character

DearWWWWWT committed:
add supports for languages other than English and Arabic

1. add supports for languages other than English and Arabic
2. code format standardized

DearWWWWWT committed:
fix logic error

nistream commented on UPMPackage/Scripts/Runtime/FastStringBuilder.cs:
```diff
@@ -66,7 +66,7 @@ public void SetValue(string text) {
 
```
请增加版权声明

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
-            }
+namespace RTLTMPro {
+  public static class LigatureFixer {
+    private static readonly List<int> LtrTextHolder = new List<int>(512);
```
_ltrTextHolder

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+namespace RTLTMPro {
+  public static class LigatureFixer {
+    private static readonly List<int> LtrTextHolder = new List<int>(512);
+    private static readonly List<int> TagTextHolder = new List<int>(512);
```
_tagTextHolder

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
-                    }
-                }
+    private static readonly HashSet<char>
+      MirroredCharsSet = new HashSet<char>(MirroredCharsMaper.MirroredCharsMap.Keys);
```
private static readonly HashSet<char> _mirroredCharsSet =
        new HashSet<char>(MirroredCharsMaper.MirroredCharsMap.Keys);
如果下一行是上一行的继续, 空4格, 不是2格

DearWWWWWT replied:
已修改

DearWWWWWT committed:
add copyright notice

DearWWWWWT committed:
update copyright

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
-                    LtrTextHolder.Add(characterAtThisIndex);
-                    continue;
+    /// <summary>
+    ///     Fixes the flow of the text.
```
对齐

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+      // This list keeps the characters that their order should be reserved and streams reserved texts into final letters.
+      LtrTextHolder.Clear();
+      TagTextHolder.Clear();
+      // import other convert function to convert character to isolate type
+      // for(int i = 0 ; i < input.Length;i++)
+      // {
+      //     input.Set(i,ArabicTable.ArabicMapper.Convert(input.Get(i)));
+      // }
```
删除

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+    /// <summary>
+    ///     Fixes the flow of the text.
+    /// </summary>
+    public static void Fix(FastStringBuilder input, FastStringBuilder output, bool farsi, bool fixTextTags,
```
每行不超过100字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+    ///     Fixes the flow of the text.
+    /// </summary>
+    public static void Fix(FastStringBuilder input, FastStringBuilder output, bool farsi, bool fixTextTags,
+      bool preserveNumbers) {
```
继续上一行空4格

DearWWWWWT replied:
已修改

DearWWWWWT committed:
code style fix

DearWWWWWT committed:
code style fix

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+            bool isValidTag = false;
+            int nextI = i;
+
+            // TagTextHolder is a List for hold on the tag content sequence
+            // Tag content's final sequence is same with arabic text
+            // in this whole Process,Tag content will be revert sequence in last unit
+            // so in this unit,Tag content's sequence is same with English
```
请重新调整, 或用中文书写

DearWWWWWT replied:
已调整

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+              TagTextHolder.Add(jChar);
+
+              if (jChar == '<') {
+                // TODO: Tag Valid judgment is to simple
+                // other invalid condition also need consider 
```
请重新调整, 或用中文书写

DearWWWWWT replied:
已调整

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
-                    output.Append(characterAtThisIndex);
-                }
+            if (isValidTag) {
+              // if a Tag is find and the Tag content is valid.
+              // fixer going to a new split.
+              // first thing is pushing the LTR list buffer into output buffer
+              // then get Tag content's reverse text and push into output buffer
+              // if the Tag content is invalid
+              // fixer continue work in old split
+              // Bug: if tag is between two English text
+              // this tag push process shuffle English text 
```
请重新调整, 或用中文书写

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+﻿// The SeedV Lab (Beijing SeedV Technology Co., Ltd.) modifications: Copyright 2024 The SeedV Lab
+// (Beijing SeedV Technology Co., Ltd.) All Rights Reserved. The modifications in this file are the intellectual property of the SeedV Lab.
```
每行做多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+    public static (ContextType[], ContextType) CharactersTypeDetermination(FastStringBuilder input) {
+      bool hasRightToLeft = false;
+      bool hasLeftToRight = false;
+      ContextType[] isRtl = new ContextType[input.Length];
```
```suggestion
     var isRtl = new ContextType[input.Length];
```

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+
+      #endregion
+
+      // if no RightToLeft and LeftToRight character is find, set to LeftToRight
```
```suggestion
      // If no RightToLeft and LeftToRight character is found, set to LeftToRight
```

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+
+      // if no RightToLeft and LeftToRight character is find, set to LeftToRight
+      if (!hasRightToLeft && !hasLeftToRight)
+        return (Enumerable.Repeat(ContextType.LeftToRight, input.Length).ToArray(), ContextType.LeftToRight);
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+                    }
+                  }
+
+                  // If the current character is the last in the input, only consider the previous type.
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+                  if (mirrorPreviousType == ContextType.Default) {
+                    mirrorPreviousType = mirrorBehindType;
+                  }
+                  // If both previous and next types are default, assume the text contains no letters.
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+                    break;
+                  }
+
+                  if (previousType == ContextType.LeftToRight && behindType == ContextType.LeftToRight &&
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+    /// </summary>
+    public static void Fix(FastStringBuilder input, FastStringBuilder output, bool farsi, bool fixTextTags,
+      bool preserveNumbers) {
+      // Some texts like tags, English words and numbers need to be displayed in their original order.
+      // This list keeps the characters that their order should be reserved and streams reserved texts into final letters.
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+          bool isAfterSymbol = Char32Utils.IsSymbol(previousCharacter);
+          bool isBeforeSymbol = Char32Utils.IsSymbol(nextCharacter);
+
+          // For cases where english words and farsi/arabic are mixed. This allows for using farsi/arabic, english and numbers in one sentence.
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+          }
+        }
+
+        if (Char32Utils.IsLetter(characterAtThisIndex) && !Char32Utils.IsRTLCharacter(characterAtThisIndex) ||
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/LigatureFixer.cs:
```diff
@@ -1,190 +1,197 @@
+          continue;
         }
+
+        // handle surrogates
```
```suggestion
        // Handle surrogates
```

DearWWWWWT replied:
已修改

nistream reviewed (changes requested):
As in comments.

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+
+                // If right character is rightest, case previous type only
+                if (i + j == input.Length - 1) pairedBehindType = pairedPreviousType;
+                // If previous type is default, there is no letter in or front this mirror character
+                if (pairedPreviousType == ContextType.Default) pairedPreviousType = pairedBehindType;
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/MixedTypographer.cs:
```diff
@@ -0,0 +1,241 @@
+                  break;
+                }
+
+                if (previousType == ContextType.LeftToRight && behindType == ContextType.LeftToRight &&
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/Types/MirroredCharsMaper.cs:
```diff
@@ -0,0 +1,52 @@
+﻿using System.Collections.Generic;
+// The SeedV Lab (Beijing SeedV Technology Co., Ltd.) modifications: Copyright 2024 The SeedV Lab
```
每行最多 100 字

DearWWWWWT replied:
已修改

nistream commented on UPMPackage/Scripts/Runtime/Types/ContextType.cs:
```diff
@@ -0,0 +1,10 @@
+﻿// The SeedV Lab (Beijing SeedV Technology Co., Ltd.) modifications: Copyright 2024 The SeedV Lab
+// (Beijing SeedV Technology Co., Ltd.) All Rights Reserved. 
```
每行最多 100 字

DearWWWWWT replied:
已修改

DearWWWWWT committed:
fix code style

DearWWWWWT committed:
delete copyright notice

DearWWWWWT commented:
已删除版权声明

nistream reviewed (approved):
LGTM
